File: src/arc-shapes/draw-comet-bevelled-tail-path.js
import { radiansToDegree, drawArcs } from "../utils/utils";
import { getSpiralArcLocus } from "../utils/formulas";
import { PathCommand } from "../utils/path-command";

/**
 * Draw top-bevelled-iontail, top-bevelled-antitail,
 * bottom-bevelled-iontail and bottom-bevelled-antitail
 * shape.
 * 
 * @param {number} radius 
 * @param {number} width 
 * @param {number} margin 
 * @param {boolean} morphingShape 
 * @param {string} shape 
 * @param {number} startingDegree 
 * @param {number} endDegree 
 * @returns {string} SVG path
 */
export default function drawCometBevelledTailPath(radius, width, margin, morphingShape, shape, startingDegree, endDegree) {
    const gap = radiansToDegree(width / radius);

    let arc1, arc2;
    if (shape === ARCSHAPE.TOP_BEVELLED_IONTAIL) {
        arc1 = getSpiralArcLocus(1, radius, width, margin, BEHAVIOR.GROW, startingDegree, endDegree - gap);
        arc2 = getSpiralArcLocus(3, radius, width, margin, BEHAVIOR.GROW, endDegree, startingDegree);
    } else if (shape === ARCSHAPE.TOP_BEVELLED_ANTITAIL) {
        arc1 = getSpiralArcLocus(1, radius, width, margin, BEHAVIOR.SHRINK, startingDegree + gap, endDegree);
        arc2 = getSpiralArcLocus(3, radius, width, margin, BEHAVIOR.SHRINK, endDegree, startingDegree);
    } else if (shape === ARCSHAPE.BOTTOM_BEVELLED_IONTAIL) {
        arc1 = getSpiralArcLocus(1, radius, width, margin, BEHAVIOR.GROW, startingDegree, endDegree);
        arc2 = getSpiralArcLocus(3, radius, width, margin, BEHAVIOR.GROW, endDegree - gap, startingDegree);
    } else if (shape === ARCSHAPE.BOTTOM_BEVELLED_ANTITAIL) {
        arc1 = getSpiralArcLocus(1, radius, width, margin, BEHAVIOR.SHRINK, startingDegree, endDegree);
        arc2 = getSpiralArcLocus(3, radius, width, margin, BEHAVIOR.SHRINK, endDegree, startingDegree + gap);
    }

    const arc1Start = arc1[0];
    const arc1End = arc1[arc1.length - 1];

    const arc2Start = arc2[0];
    const arc2End = arc2[arc2.length - 1];

    /*
        Draw path pattern:
        without morph           : M -> A1 -> L  ->       A2
        without morph (reverse) : M -> A1 ->             A2 -> Z
        with morph              : M -> A1 -> C1 -> C2 -> A2 -> C3 -> C4
    */
    let M, A1, LC1, C2, A2, ZC3, C4;

    M = PathCommand.moveTo.M(arc1Start.point.x, arc1Start.point.y);
    A1 = drawArcs(morphingShape, 0, 0, arc1.slice(1));
    A2 = drawArcs(morphingShape, 0, 1, arc2.slice(1));

    if (morphingShape) {
        LC1 = PathCommand.cubicBezierCurve.C(arc1End.point.x, arc1End.point.y, arc1End.point.x, arc1End.point.y, arc1End.point.x, arc1End.point.y);

        ZC3 = PathCommand.cubicBezierCurve.C(arc2End.point.x, arc2End.point.y, arc2End.point.x, arc2End.point.y, arc2End.point.x, arc2End.point.y);

        C4 = PathCommand.cubicBezierCurve.C(arc2End.point.x, arc2End.point.y, arc1Start.point.x, arc1Start.point.y, arc1Start.point.x, arc1Start.point.y);

        if (shape === ARCSHAPE.TOP_BEVELLED_IONTAIL || shape === ARCSHAPE.BOTTOM_BEVELLED_IONTAIL) {
            C2 = PathCommand.cubicBezierCurve.C(arc1End.point.x, arc1End.point.y, arc2Start.point.x, arc2Start.point.y, arc2Start.point.x, arc2Start.point.y);
        } else if (shape === ARCSHAPE.TOP_BEVELLED_ANTITAIL || shape === ARCSHAPE.BOTTOM_BEVELLED_ANTITAIL) {
            C2 = PathCommand.cubicBezierCurve.C(arc1End.point.x, arc1End.point.y, arc1End.point.x, arc1End.point.y, arc1End.point.x, arc1End.point.y);
        }

        return `${M} ${A1} ${LC1} ${C2} ${A2} ${ZC3} ${C4}`;
    } else {
        if (shape === ARCSHAPE.TOP_BEVELLED_IONTAIL || shape === ARCSHAPE.BOTTOM_BEVELLED_IONTAIL) {
            LC1 = PathCommand.lineTo.L(arc2Start.point.x, arc2Start.point.y);

            return `${M} ${A1} ${LC1} ${A2}`;
        } else if (shape === ARCSHAPE.TOP_BEVELLED_ANTITAIL || shape === ARCSHAPE.BOTTOM_BEVELLED_ANTITAIL) {
            ZC3 = PathCommand.closePath;

            return `${M} ${A1} ${A2} ${ZC3}`;
        }

    }

}

File: wasm/src/arc_shape/draw_comet_bevelled_tail_path.rs
use crate::util::{
    constant::ArcShape,
    draw_arcs,
    formula::{get_spiral_arc_locus, Behavior, SpiralArcLayer},
    path_command, radians_to_degree,
};

#[allow(non_snake_case)]
pub fn draw_comet_bevelled_tail_path(
    radius: f64,
    width: f64,
    margin: f64,
    morphing_shape: bool,
    shape: ArcShape,
    starting_degree: f64,
    end_degree: f64,
) -> String {
    let gap = radians_to_degree(width / radius);

    let (behavior, arc1_degree, arc2_degree) = match shape {
        ArcShape::TopBevelledIontail => {
            let arc1 = (starting_degree, end_degree - gap);
            let arc2 = (end_degree, starting_degree);

            (Behavior::Grow, arc1, arc2)
        }
        ArcShape::TopBevelledAntitail => {
            let arc1 = (starting_degree + gap, end_degree);
            let arc2 = (end_degree, starting_degree);

            (Behavior::Shrink, arc1, arc2)
        }
        ArcShape::BottomBevelledIontail => {
            let arc1 = (starting_degree, end_degree);
            let arc2 = (end_degree - gap, starting_degree);

            (Behavior::Grow, arc1, arc2)
        }
        ArcShape::BottomBevelledAntitail => {
            let arc1 = (starting_degree, end_degree);
            let arc2 = (end_degree, starting_degree + gap);

            (Behavior::Shrink, arc1, arc2)
        }
        _ => panic!("Shape not allowed"),
    };

    let arc1 = get_spiral_arc_locus(
        SpiralArcLayer::L1,
        radius,
        width,
        margin,
        behavior,
        arc1_degree.0,
        arc1_degree.1,
    );
    let arc1_start = arc1[0];
    let arc1_end = arc1[arc1.len() - 1];

    let arc2 = get_spiral_arc_locus(
        SpiralArcLayer::L3,
        radius,
        width,
        margin,
        behavior,
        arc2_degree.0,
        arc2_degree.1,
    );
    let arc2_start = arc2[0];
    let arc2_end = arc2[arc2.len() - 1];

    /*
        Draw path pattern:
        without morph           : M -> A1 -> L  ->       A2
        without morph (reverse) : M -> A1 ->             A2 -> Z
        with morph              : M -> A1 -> C1 -> C2 -> A2 -> C3 -> C4
    */
    let M = path_command::MoveTo::M(arc1_start.point.x, arc1_start.point.y);
    let A1 = draw_arcs(morphing_shape, 0, 0, &arc1[1..]);
    let A2 = draw_arcs(morphing_shape, 0, 1, &arc2[1..]);

    if morphing_shape {
        let LC1 = path_command::CubicBezierCurve::c(
            arc1_end.point.x,
            arc1_end.point.y,
            arc1_end.point.x,
            arc1_end.point.y,
            arc1_end.point.x,
            arc1_end.point.y,
        );

        let ZC3 = path_command::CubicBezierCurve::C(
            arc2_end.point.x,
            arc2_end.point.y,
            arc2_end.point.x,
            arc2_end.point.y,
            arc2_end.point.x,
            arc2_end.point.y,
        );

        let C4 = path_command::CubicBezierCurve::C(
            arc2_end.point.x,
            arc2_end.point.y,
            arc1_start.point.x,
            arc1_start.point.y,
            arc1_start.point.x,
            arc1_start.point.y,
        );

        let C2 = match shape {
            ArcShape::TopBevelledIontail | ArcShape::BottomBevelledIontail => {
                path_command::CubicBezierCurve::C(
                    arc1_end.point.x,
                    arc1_end.point.y,
                    arc2_start.point.x,
                    arc2_start.point.y,
                    arc2_start.point.x,
                    arc2_start.point.y,
                )
            }
            ArcShape::TopBevelledAntitail | ArcShape::BottomBevelledAntitail => {
                path_command::CubicBezierCurve::C(
                    arc1_end.point.x,
                    arc1_end.point.y,
                    arc1_end.point.x,
                    arc1_end.point.y,
                    arc1_end.point.x,
                    arc1_end.point.y,
                )
            }
            _ => panic!("Shape not allowed"),
        };

        format!("{M} {A1} {LC1} {C2} {A2} {ZC3} {C4}")
    } else {
        match shape {
            ArcShape::TopBevelledIontail | ArcShape::BottomBevelledIontail => {
                let LC1 = path_command::LineTo::L(arc2_start.point.x, arc2_start.point.y);

                format!("{M} {A1} {LC1} {A2}")
            }
            ArcShape::TopBevelledAntitail | ArcShape::BottomBevelledAntitail => {
                let ZC3 = path_command::CLOSE_PATH;

                format!("{M} {A1} {A2} {ZC3}")
            }
            _ => panic!("Shape not allowed"),
        }
    }
}
